.circleci/parse_yaml.py

```diff
@@ -0,0 +1,273 @@
+    """
+    from tqdm import tqdm
+    import requests
+    import math
+    
+    r = requests.get(path, stream=True)
+    total_size = int(r.headers.get('content-length', 0))
+    block_size = 1024
+    wrote = 0
+    file_name = path.split('/')[-1]
+    file_path = outdir + file_name
+    
+    with open(file_path, 'wb') as f:
+        for data in tqdm(r.iter_content(block_size), 
+                         total=math.ceil(total_size//block_size), 
+                         unit='KB', 
+                         leave=True,
+                         ncols=42,
+                         unit_scale=True,
+                         unit_divisor=1024):
+            wrote = wrote + len(data)
+            f.write(data)
+    
+    if total_size != 0 and wrote != total_size:
+        logging.error(f' check the file link and try again.')
+        sys.exit(1)
+
+
+def validate_hash(path, hash):
+    """Validate a file against an MD5 hash value."""
+    import hashlib
+
+    md5 = hashlib.md5()
+    
+    with open(path, 'rb') as f:
+        while True:
+            chunk = f.read(10000000)
+            if not chunk:
+                break
+            md5.update(chunk)
+    
+    return md5.hexdigest() == hash
+
+
+class Yamler:
+    """A collection of utilities to parse the yaml file, check metadata
+    and trigger cimr processing of the contributed file
+    """
+
+    def __init__(self, yaml_data):
+        self.yaml_data = yaml_data
+        self.data_type = None
+        self.keys = None
+        self.hash = None
+        self.outdir = None
+        self.downloaded_file = None
+
+
+    def pick_keys(self):
+        """List keys for the dictionarized yaml data and store in self.
+        The following keys are expected:
+        ['defined_as', 'data_file', 'contributor', 'data_info', 'method']
+        """
+        self.keys = self.yaml_data.keys()
+
+
+    def set_data_type(self):
+        """One of the following data_type variables are expected:
+        ['gwas', 'twas', 'eqtl', 'sqtl', 'pqtl', 'tad']
+        """
+        if self.yaml_data['data_info']['data_type'] is not None:
+            self.data_type = self.yaml_data['data_info']['data_type']
+        else:
+            logging.error(f' there is no data_type indicated.')
+            sys.exit(1)
+
+
+    def download(self):
+        """Check if provided weblink to the file exists. 
+        Download if verified.
+        """
+        path = self.yaml_data['data_file']['location']['url']
+        self.downloaded_file = path.split('/')[-1]
+
+        outdir_root = 'submitted_data/'
+        pathlib.Path(outdir_root).mkdir(exist_ok=True)
+        self.outdir = outdir_root + str(self.data_type) + '/'
+        pathlib.Path(self.outdir).mkdir(exist_ok=True)
+
+        if verify_weblink(path):
+            logging.info(f' starting download')
+            download_file(path, self.outdir)
+            self.hash = self.yaml_data['data_file']['location']['md5']
```

Comment:
My failed test indicates that the following line should be added:
```python
self.downloaded_file = self.outdir + self.downloaded_file
```
otherwise `check_hash()` would fail because the downloaded file cannot be found. (This is exactly why the `except` block should be moved up a little bit, as I mentioned earlier.)

Reply:
good catch. thanks.